### DISCUSSION TITLE
[General] Adding more VMs into my private testnet

### DISCUSSION THREAD

**Discussion by ghost** [zkSync-Community-Hub/zksync-developers#688] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

Hi everyone,
I'm currently running both L1 and L2 nodes using the official [zk-chains-docker-compose.yml](https://github.com/matter-labs/local-setup/blob/main/zk-chains-docker-compose.yml) setup without any modifications, and everything works perfectly. Now, I want to set up a second instance of both L1 and L2 nodes on separate virtual machines (VMs).
Could someone guide me on the best way to achieve this? Specifically, I'm curious about:

1. What configuration changes (if any) are needed in the docker-compose files for each VM?
2. How do I avoid conflicts between the two instances (ports, data directories, etc.)?
3. Should I be making any modifications to the network or communication between these two setups?
    
Any detailed instructions or advice from those who've done this would be greatly appreciated!
Thanks in advance!

**Comment by alhenaai** [comment]
Hello,

Setting up multiple instances of L1 and L2 nodes on separate virtual machines using the Dockerized setup from ZKsync involves a few key configuration adjustments to ensure smooth operation and avoid conflicts. Here’s a general guide on how to proceed:

1. **Clone the Repository**: Ensure each VM has its own copy of the repository. You can clone the ZKsync project repository using:
   ```
   git clone https://github.com/matter-labs/local-setup
   ```

2. **Modify Docker Compose Files**:
   - **Ports**: Change the ports in the `docker-compose.yml` file to avoid conflicts. For example, if the default L2 RPC port is 3050 on the first VM, you might set it to 3150 on the second VM.
   - **Data Directories**: Modify the volume bindings to use different directories for each instance to prevent data overlap.

3. **Network Configuration**:
   - Ensure that the network settings in the Docker compose file allow for inter-VM communication if the nodes need to interact.
   - Adjust the `ETH_CLIENT_WEB3_URL` environment variable if your L1 nodes are supposed to communicate with an external Ethereum node or with each other.

4. **Launching the Nodes**:
   - Navigate to the directory containing the modified `docker-compose.yml` file.
   - Start the nodes using:
     ```
     docker-compose up
     ```

5. **Testing and Validation**:
   - Verify that each node operates correctly and can connect to its counterpart on the other VM.
   - Check the logs for any error messages or conflicts.

For detailed instructions on setting up and configuring the Docker environment, you can refer to the [Docker L1 - L2 Nodes documentation](https://docs.zksync.io/build/test-and-debug/dockerized-l1-l2-nodes).

If you encounter specific issues or need further customization, consider adjusting the network IDs and ensuring that each instance has a unique configuration to operate independently or in tandem, depending on your testing requirements.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by ghost** [reply]
Need more details with examples

**Reply by alhenaai** [reply]
Certainly! Let's dive into more detailed steps with examples for setting up multiple instances of L1 and L2 nodes on separate virtual machines using Docker:

### Step 1: Clone the Repository
On each VM, clone the repository to ensure each has its own separate setup:
```bash
git clone https://github.com/matter-labs/local-setup.git
cd local-setup
```

### Step 2: Modify Docker Compose Files
Open the `zk-chains-docker-compose.yml` file in an editor. You’ll need to make the following changes:

#### Ports
Change the ports to avoid conflicts. For example, if the original file has:
```yaml
ports:
  - "3050:3050" # L2 RPC port
  - "8545:8545" # L1 RPC port
```
Change it on the second VM to:
```yaml
ports:
  - "3150:3050" # L2 RPC port
  - "8645:8545" # L1 RPC port
```

#### Data Directories
Change the volume bindings to use different directories. Original:
```yaml
volumes:
  - ./data/l2:/zksync/data
```
Modified on the second VM:
```yaml
volumes:
  - ./data_vm2/l2:/zksync/data
```

### Step 3: Network Configuration
Ensure the VMs can communicate if needed. Modify the `ETH_CLIENT_WEB3_URL` to point to the correct Ethereum node or another L1 node. For example:
```yaml
environment:
  ETH_CLIENT_WEB3_URL: "http://192.168.1.100:8545" # IP of the first VM if they need to interact
```

### Step 4: Launching the Nodes
Navigate to the directory with the `docker-compose.yml` file and start the nodes:
```bash
docker-compose up -d
```

### Step 5: Testing and Validation
- Check that the nodes are running and accessible on the defined ports using:
  ```bash
  curl http://localhost:3150 # Adjust port for the specific VM
  ```
- Review the logs for any errors:
  ```bash
  docker-compose logs -f
  ```

### Additional Tips
- **Ensure Isolation**: If the nodes do not need to interact, keep their network settings isolated to prevent any unintended interference.
- **Security**: Never expose test nodes with open ports to the public internet without proper security measures.

By following these steps, you can successfully run multiple instances of L1 and L2 nodes on separate VMs, each with its own configuration and data storage. This setup is ideal for testing scalability and interaction between different nodes in a controlled environment.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
Is it correct that you want to run another setup on a separate VM that won't have the same chain IDs as the one that you have now and you want both setups to be able to communicate with each other? 

What is your end goal with this so that we can help you better? Are you just playing around or wanting to see what it would take to have multiple chains running and all working together?

**Reply by ghost** [reply]
No, my new VM running reth and ZK should be able to connect to the main VM which runs the official [zk-chains-docker-compose.yml](https://github.com/matter-labs/local-setup/blob/main/zk-chains-docker-compose.yml). Chain IDs should remain the same and the nodes should sync and work with my main VM, effectively acting as a real L1/L2 testnet.
But I don't want to deploy all other components, like PostgreSQL database, HyperExplorer, L1, L2 explorers, Blockscout, data fetchers and workers. For those, I would use my Main VM

**Reply by bxpana** [reply, marked as answer]
Are you looking to have the VM that runs the zk-chains-docker-compose.yml be the only one that can write and the other "nodes" you spin up could read and send transactions to the main VM? 

Regardless, I did check with the team and right now there is no one saying it's not possible, but I don't think we've tried yet or documented any steps on how you would do that.

**Comment by popzxc** [comment]
@YeagerAI-Bulat not sure if it answers your question, but if you are don't necessarily have to run L2 nodes in docker containers, you can use [zk_toolbox](https://github.com/matter-labs/zksync-era/tree/main/zk_toolbox). It's capable of setting up several L2s connected to a single L1 instance, with the ability to switch between chains, run them simultaneously, and interact. It also shares resources where possible (e.g. it uses a single Postgres container to store all the tables) and manages configurations, e.g. so that ports don't collide. Finally, you can run external nodes specific to the chain of your choice as well.

However, if you _have_ to run L2 nodes in docker containers, then as @bxpana said, it's possible, but we don't have ready guides for that. Probably, the best way to approach it would be to learn the configuration format for L2 node, and then prepare the required configs for chains.
